### PR TITLE
Add diffusion model implementation

### PR DESCRIPTION
**Pull request by lesterphillip**
Hi, I'm adding a PR for a basic diffusion model implementation.

To test it, you can simply replace the configuration file in run.sh using the one in `conf/diffusion.yaml`

Let me know if you encounter any bugs or have any suggestions for revision.

### CODE REVIEW

File: s3prl_vc/bin/decode.py [hunk 1]
```diff
@@ -245,12 +257,20 @@ def main():
...
-            out = outs[0]
+
+            if config["model_type"] == "Taco2_AR":
+                outs, _ = model(hs, hlens, spk_embs=spemb, f0s=f0s)
+                out = outs[0]
+
+            elif config["model_type"] == "Diffusion":
+                out = model.inference(hs, spk=spemb)
+                # inverse normalization
+                out = config["trg_stats"]["mean"] + (out * config["trg_stats"]["scale"])
```
Comment (unilight) [inline s3prl_vc/bin/decode.py]: Is it possible to move this inside the inference function? (Of course this will also affect the signature of the model class above)

File: s3prl_vc/bin/decode.py [hunk 2]
```diff
@@ -194,17 +194,29 @@ def main():
...
-        config["num_mels"],
-        config["sampling_rate"]
-        / config["hop_size"]
-        * upstream_featurizer.downsample_rate
-        / 16000,
-        config["trg_stats"],
-        use_spemb=config.get("use_spk_emb", False),
-        **config["model_params"],
-    ).to(device)
+    if config["model_type"] == "Taco2_AR":
```
Comment (unilight) [inline s3prl_vc/bin/decode.py]: Wanted to avoid writing these if statements here (or else it would be meaningless to use the model_class variable here). Please try to keep a minimum number of common arguments shared among different models, and leave the rest to named arguments. I know that when originally designing the signature, I could not see the need of implementing the diffusion model coming, so there are definitely some parameters not essentially needed or used by the diffusion model. Please modify the original model signature if needed.

File: s3prl_vc/bin/decode.py [hunk 3]
```diff
@@ -245,12 +257,20 @@ def main():
...
                 spemb = torch.from_numpy(spemb).unsqueeze(0).float().to(device)
             ilens = torch.LongTensor([x.shape[0]]).to(device)
-
             start_time = time.time()
             all_hs, all_hlens = upstream_model(xs, ilens)
             hs, hlens = upstream_featurizer(all_hs, all_hlens)
-            outs, _ = model(hs, hlens, spk_embs=spemb, f0s=f0s)
-            out = outs[0]
+
+            if config["model_type"] == "Taco2_AR":
```
Comment (unilight) [inline s3prl_vc/bin/decode.py]: Same as above.

File: s3prl_vc/bin/train.py [hunk 4]
```diff
@@ -290,9 +320,14 @@ def _eval_epoch(self):
...
-                self._genearete_and_save_intermediate_result(batch)
-            else:
+            continue
+            if self.config["model_type"] == "Taco2_AR":
+                # normalize
+                if eval_steps_per_epoch == 1:
+                    self._genearete_and_save_intermediate_result(batch)
+                else:
+                    continue
+            elif self.config["model_type"] == "Diffusion":
```
Comment (unilight) [inline s3prl_vc/bin/train.py]: Does this mean that no dev samples are generated for the diffusion model? Is there a specific reason?

File: s3prl_vc/bin/train.py [hunk 5]
```diff
@@ -735,17 +770,29 @@ def main():
...
-        config["sampling_rate"]
-        / config["hop_size"]
-        * upstream_featurizer.downsample_rate
-        / 16000,
-        config["trg_stats"],
-        use_spemb=config.get("use_spk_emb", False),
-        **config["model_params"],
-    ).to(device)
+
+    if config["model_type"] == "Taco2_AR":
```
Comment (unilight) [inline s3prl_vc/bin/train.py]: Same in train.py

File: s3prl_vc/models/__init__.py [hunk 6]
```diff
@@ -1 +1,4 @@
 from .Taco2_AR import *  # NOQA
+from .diffsinger.denoiser import * # NOQA
```
Comment (unilight) [inline s3prl_vc/models/__init__.py]: I think this line is not needed?

File: s3prl_vc/models/__init__.py [hunk 7]
```diff
@@ -1 +1,4 @@
...
+from .diffsinger.diffusion import * # NOQA
```
Comment (unilight) [inline s3prl_vc/models/__init__.py]: I think this line is not needed, same as above?

File: README.md [hunk 8]
```diff
@@ -71,4 +71,6 @@ This repo is greatly inspired by the following repos. Or I should say, many code
 
 - [ESPNet](https://github.com/espnet/espnet)
 - [S3PRL](https://github.com/s3prl/s3prl)
-- [ParallelWaveGAN](https://github.com/kan-bayashi/ParallelWaveGAN/)
\ No newline at end of file
+- [ParallelWaveGAN](https://github.com/kan-bayashi/ParallelWaveGAN/)
+- [NNSVS](https://github.com/nnsvs/nnsvs/)
+- [DiffSinger](https://github.com/MoonInTheRiver/DiffSinger)
```
Comment (unilight) [inline README.md:76]: Not really in a hurry but probably can write a small section on this upgrade.

File: s3prl_vc/losses/l1_loss.py [hunk 9]
```diff
@@ -16,9 +16,12 @@ class L1Loss(torch.nn.Module):
                               (2) masked loss calculation
     """
 
-    def __init__(self):
+    def __init__(self, loss_func="l1"):
         super(L1Loss, self).__init__()
-        self.objective = torch.nn.L1Loss(reduction="mean")
+        if loss_func == "l1":
+            self.objective = torch.nn.L1Loss(reduction="mean")
+        elif loss_func == "mse":
```
Comment (unilight) [inline s3prl_vc/losses/l1_loss.py]: Feels a bit strange to offer an L2 loss option in the l1 loss class...

File: s3prl_vc/models/diffsinger/__init__.py [hunk 10]
```diff
@@ -0,0 +1,2 @@
+from .denoiser import DiffNet
```
Comment (unilight) [inline s3prl_vc/models/diffsinger/__init__.py:1]: I wonder if this file is really needed?

File: s3prl_vc/models/diffusion.py [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+            y_mel = y_mel[:, : x.size(1), :]
+
+        if spk is not None:
+            spk = spk.squeeze(-1)
+
+        mel_inp = x
+        mel_ = self.mel_model(mel_inp, lengths, y_mel, spk)
+        return (mel_[0], mel_[1], lengths)
+
+    def inference(self, x, spk=None):
```
Comment (unilight) [inline s3prl_vc/models/diffusion.py]: Somehow thinking we either (1) also write a `inference` function in `Taco2_AR.py` and unify the signature, or (2) avoid writing `inference` here by, for example, decide whether we are in the inference mode by whether `y_mel` (which is `targets` in Taco2_AR.py?) is given.

**Review by unilight — CHANGES_REQUESTED** [review]

**Comment by unilight** [comment]
Thanks Lester, really nice work! For now, I think we just need to unify some code structure and style. I'll try to test this model on other datasets (VCC2020 for instance) later.

**Comment by lesterphillip** [comment]
Thanks for checking the code! I've made some changes but here's a summary:

- I made the model signatures for both Taco2AR and Diffusion to be the same to remove most of the if conditions in the train and decode scripts.
- I transferred the feature normalization inside the model class (for both Taco2 and Diffusion) instead of the training loop.

Let me know if there are other changes needed!

File: s3prl_vc/models/__init__.py (hunk 7 above)
Comment (lesterphillip) [reply s3prl_vc/models/__init__.py]: Removed.

File: s3prl_vc/losses/l1_loss.py (hunk 9 above)
Comment (lesterphillip) [reply s3prl_vc/losses/l1_loss.py]: I separated them into two files instead.

File: s3prl_vc/models/diffusion.py (hunk 11 above)
Comment (lesterphillip) [reply s3prl_vc/models/diffusion.py]: I think it would be better to instead do the second option, which I reflected in the changes.

File: s3prl_vc/bin/train.py (hunk 5 above)
Comment (lesterphillip) [reply s3prl_vc/bin/train.py]: Should be resolved.

File: s3prl_vc/bin/train.py (hunk 4 above)
Comment (lesterphillip) [reply s3prl_vc/bin/train.py]: This is my bad, I fixed this already.

File: s3prl_vc/bin/decode.py (hunk 3 above)
Comment (lesterphillip) [reply s3prl_vc/bin/decode.py]: Should be resolved.

File: s3prl_vc/bin/decode.py (hunk 2 above)
Comment (lesterphillip) [reply s3prl_vc/bin/decode.py]: Should be resolved.

File: s3prl_vc/bin/decode.py (hunk 1 above)
Comment (lesterphillip) [reply s3prl_vc/bin/decode.py]: I moved the normalization inside the models for both Taco2 and diffusion.

File: s3prl_vc/models/__init__.py (hunk 6 above)
Comment (lesterphillip) [reply s3prl_vc/models/__init__.py]: Should be resolved.

File: s3prl_vc/models/diffsinger/__init__.py (hunk 10 above)
Comment (lesterphillip) [reply s3prl_vc/models/diffsinger/__init__.py:1]: This one is needed since we call these two classes inside the diffusion.py file.

File: s3prl_vc/bin/train.py [hunk 12]
```diff
@@ -213,19 +214,23 @@ def _train_step(self, batch):
...
-        # normalize output
-        outs = (outs - self.config["trg_stats"]["mean"]) / self.config["trg_stats"][
-            "scale"
-        ]
-        ys = (ys - self.config["trg_stats"]["mean"]) / self.config["trg_stats"]["scale"]
+        if self.config["model_type"] == "Diffusion":
+            # diffusion model needs the inputs and outputs to be
+            # of the same length, so there are cases where it is
+            # cut to match with the upsampled upstream features
+            olens = outs_lens
```
Comment (unilight) [inline s3prl_vc/bin/train.py]: Hmm, can we put this in the loss required by the diffusion model? Please refer to this file: https://github.com/unilight/s3prl-vc/blob/main/s3prl_vc/losses/l1_loss.py#L25-L30

File: s3prl_vc/bin/train.py [hunk 13]
```diff
@@ -294,6 +299,8 @@ def _eval_epoch(self):
                 self._genearete_and_save_intermediate_result(batch)
             else:
                 continue
+            #elif self.config["model_type"] == "Diffusion":
```
Comment (unilight) [inline s3prl_vc/bin/train.py]: Remember to remove this.

File: s3prl_vc/bin/train.py [hunk 14]
```diff
@@ -376,17 +383,19 @@ def _plot_and_save(
...
             out = out[:olen]
             y = y[:olen]
-            _plot_and_save(
-                out.cpu().numpy(),
-                dirname + f"/outs/{idx}_out.png",
-                ref=y.cpu().numpy(),
-                origin="lower",
-            )
+
+            if self.config["model_type"] == "Taco2_AR":
```
Comment (unilight) [inline s3prl_vc/bin/train.py]: Same as my question in the last round of review -- is there a specific reason why the diffusion model cannot generate dev set samples?

**Review by unilight — CHANGES_REQUESTED** [review]
Most of the fixes LGTM! Just a few comments left. Much thanks!!

File: s3prl_vc/bin/train.py (hunk 13 above)
Comment (lesterphillip) [reply s3prl_vc/bin/train.py]: Removed

File: s3prl_vc/bin/train.py (hunk 14 above)
Comment (lesterphillip) [reply s3prl_vc/bin/train.py]: The if condition is only for generating the plots of the attention weights, so the code should generate samples for both taco and diffusion models here.

File: s3prl_vc/bin/train.py (hunk 12 above)
Comment (lesterphillip) [reply s3prl_vc/bin/train.py]: Good idea, I moved it over to the L2 loss then.

File: s3prl_vc/bin/train.py (hunk 14 above)
Comment (unilight) [reply s3prl_vc/bin/train.py]: Wait I thought "out" is the generated mel? So both Taco2_AR and Diffusion should generate these?

**Review by unilight — CHANGES_REQUESTED** [review]
One last comment, please check!

**Review by unilight — APPROVED** [review]
